### PR TITLE
feat: add option to auto-install missing parsers

### LINKED ISSUE #2108: Automatically install just one parser on encountering a new filetype
**Summary**
I want to be able to automatically install the relevant parser when encountering a new filetype.

How could I install *only* the required parser for this new filetype?

**Is your feature request related to a problem? Please describe.**

I have PolyGlot installed.  When I open a file of a new filetype how do I know if I need to install the TS parser for it?  It would be easier to not need to check, but that the TS parser is installed automatically if it exists.

**Describe the solution you'd like**

A `.vimrc` snippet for the following:

I'm not a vim guru. I'd love to have a complete example snippet for how to add the relevant entry in `ensure_installed` upon encountering a new filetype, and then what to do to call either  `TSInstall` or `TSUpdate` as appropriate.

https://github.com/nvim-treesitter/nvim-treesitter/issues/293 says it's possible to add to:
```
ensure_installed = { 'typescript' }
```

**Describe alternatives you've considered**
I could install **all** parsers but that would be a waste of time / disk / bandwidth especially when `ssh`-ing to a new machine to edit a single text file.

### PR DESCRIPTION
**Pull request by smjonas**
Just putting this out here for further discussion. I tested it and it seems to work fine.

Closes #2108.

### CODE REVIEW

File: lua/nvim-treesitter/install.lua
Comment (kyazdani42) [inline lua/nvim-treesitter/install.lua:430]: i'd add `once = true` also

File: lua/nvim-treesitter/install.lua
Comment (smjonas) [reply lua/nvim-treesitter/install.lua:437]: I simply took this from the original code [here](https://github.com/nvim-treesitter/nvim-treesitter/issues/2108#issuecomment-995069984). Likely that this is no longer necessary.

**Review by kyazdani42 — APPROVED** [review]
let's wait for another review on this :)

File: lua/nvim-treesitter/install.lua
Comment (smjonas) [reply lua/nvim-treesitter/install.lua:436]: Yes, tested for haskell, java, perl, typescript and other filetypes that I had not installed locally.

**Comment by smjonas** [comment]
@kyazdani42 seems like `once = true` should not be added to the autocommand after all, consider this scenario:
- `:e ~/test.pl` => perl parser is installed as expected
- `:e ~/test.hs` => haskell parser is not installed because the autocommand has been removed
So I removed it.

**Comment by kyazdani42** [comment]
@smjonas indeed, i mistread the au, i though it was binding for specific filetypes.

**Review by kyazdani42 — APPROVED** [review]

**Comment by numToStr** [comment]
FYI, if you `auto_install` markdown parser you probably need both `markdown` and `markdown_inline`

**Comment by syphar** [comment]
@smjonas thank you for implementing this! came here from #2108 . 

Currently the highlight is broken directly after installation. Is there a way to fix this? 

( when I use `TSInstallSync` from the original examples in the issue, it's directly correct )

**Comment by smjonas** [comment]
@syphar For me, the highlight work immediately after the parser is installed (I verified that `reattach_module` in `install.lua` is called). By broken, do you mean only Treesitter highlighting doesn't work or is there no highlighting at all?

**Comment by syphar** [comment]
> @syphar For me, the highlight work immediately after the parser is installed (I verified that `reattach_module` in `install.lua` is called). By broken, do you mean only Treesitter highlighting doesn't work or is there no highlighting at all?

Just normal hilight. 

But when this works for you then I assume the problem is somewhere in my setup, and not in `auto_install=true`. 

Thank you!

**Pull request merged**
